### DISCUSSION TITLE
Lightweight X server

### DISCUSSION THREAD

**Discussion by vvaltchev** [vvaltchev/tilck#81] [Q&A] [answered]
@el3ctr0lyte wrote in #80:

> There are lightweight x server implementations like Xfree86 and Nano-x that could maybe be added as extra stuff that you have to configure with the build scripts. It would be impressive to see how well this would work and what applications would run. Although i think you would probably also need gtk and qt5 which would bloat it a bit. Just something to think about.

**Comment by vvaltchev** [comment, marked as answer]
Hi! Thanks for taking interest in Tilck.

About your question: Tilck does **not** aim at becoming a desktop operating system at all, nor at being 100% compatible with Linux. Nevertheless, with time, more and more Linux syscalls will be supported as use cases come out and I have time to invest in development.

Running an X server would require *many* IPC syscalls like sockets that are not supported by Tilck, at the moment. Briefly, the project is far from there and other features have a higher priority. At some point, sockets and unix domain sockets will be supported, and, with some additional effort it will be possible to run an X server on it, but that's a _side work_, almost like running `vim` and `fbDOOM`: it's cool being able to do that, but it's **not the primary goal of this operating system**.

I've heard about **Nano-x** and, one day, that might be a candidate for Tilck, certainly not the XFree86 beast :-)

BTW, with "port" I mean:

- Compile the project for i686-linux with the GCC musl toolchain I use and use static linking
- Run the project on Tilck using a syscall tracer
- Examine the syscalls it tries to perform
- Implement all of those syscalls in Tilck and repeat the process until the project runs as expected

In **conclusion,** _sorry to disappoint_ but there won't be X support any time soon, because the lack of required features and because running X is _completely outside_ project's goals.

**Reply by mung-kie** [reply]
HAve you implement sockets yet?

What is simplest compile options toolchain etc for tilck musl static build?

If you have details I may try have a try this evening on a linuxmint VM(I have not tried anything with tilck yet) please give the idiots guide and I will sent you copy of source/binary and any compile logs.

**Reply by vvaltchev** [reply]
Hello @mung-kie, sorry but Tilck does not have support for UNIX sockets yet. Make any existing X server to work on Tilck will require an insane amount of effort, plus deep expertise in the project itself. Definitively, not a task that could be done in a few days.

If you want to help, please take a look at the list of issues for new contributors.

**Reply by mung-kie** [reply]
I was not intending to run a standard X11 server, I was going to compile a minimal X server and see what was missing from tilck and maybe hack some work around for IPC.

I compiled a QEMU image and run it on vmware (qemu-img convert -f raw -O vmdk tilck.img tilck. vmdk) but there is no /dev/mouse and also I cannot see how to add program into 'build/tilck.img' when run 'make rem' ?

Any advice welcome, I am afraid I cannot really help with your issues as I am not really a kernel developer, I doubt I will go much further with this unless it looks easy.

<img width="401" alt="tilckSS" src="https://user-images.githubusercontent.com/114813840/193648751-cffa4eb4-3ce8-43c5-9336-2b236b05d88d.png">

**Reply by vvaltchev** [reply]
Hello @mung-kie, I appreciate your enthusiasm so much. I'm impressed that, despite my previous answer, you persisted and found a way to build and run Tilck. 

The problem is that making any X server to work on Tilck will require an _insane_ amount of work. I can guarantee you that, as I've written so far everything in this project (and ported smaller projects like vim and fbDOOM) on it. For example, you don't see /dev/mouse because I haven't implemented any mouse drivers so far. Also, the only IPC that Tilck currently support is anonymous pipes. That's definitively not enough for any X server. At least UNIX domain sockets will be required, I guess, let alone a real network stack for TCP/IP sockets. Also, there is no support for sharing memory across processes yet.

Porting any X server to Tilck will require a deep understanding of the whole project, and probably a few months of (full-time) work to implement the required features. It might the case that making Wayland to work on Tilck could be easier, not sure. Therefore, I cannot somehow guide you at the moment in this monumental task, but I can offer all of my help to introduce you to the project and start with smaller and simpler tasks. Over time, you'll learn enough that you'd be able to take a project like that, if you really want. I'll be there for you in that case. But, I cannot offer you "false hopes" that with a bit of hacking here and there an X server will run on Tilck.

Still, I won't skip answering to your questions. To copy something into the Tilck's image, drop the file in `sysroot` directory and then run `make rem`. That's not the proper way to do so of course, but it's perfect for debugging and making experiements.

In general, the way to experiment with new ports is:
  - build the application with the gcc musl toolchain that Tilck uses and link statically.  E.g. `file <elf binary>` must report "statically linked". If it's a dynamic executable, even if statically linked, it won't run on Tilck.

 - copy the app in to sysroot, rebuild the image and run ./build/run_qemu. It's much more easier to do everything with QEMU as I have plenty of scripts for that. 

 - After you got the QEMU window opened, click on `view` and open the serial0 window. Then detach the window and return to the main video on the first window. Now you'll have two terminals with Tilck, one on video, the other via serial port. Use the one connected via (virtual) serial port to run `tracer`. Press `h` to get help, `p` to get the list of processes, `t` to enter the list of PIDs to trace, by following the children implicitly. With ENTER you can start or stop the actual tracing. This way, you'll have an idea of what the application is doing in real time. Often, it's convenient for me to just trace the ASH shell on TTY 1 and then run the application, so that the tracing reports everything from the start.

The tracing infrastructure is not perfect, and many argument types are not "rendered" yet, so they are shown in raw form, but still, it help a lot. Especially because you can filter which syscalls should be traced. I was able to debug both `vim` and `fbDOOM` this way. You can see how to do that in a slightly different way also in this video from 2 years ago: https://youtu.be/8wrt3RHefhI?t=620

I think that's all you need at least for the moment to have plenty of fun with this project and observe how new apps will behave on it.

**Comment by mung-kie** [comment]
I am afraid I got a bit busy and doubt I can do anything of use, so appologise as I may have wasted your time.
I managed to find some time this evening so have compiled the Xfbdev kdrive server using the tilck toolchain.

But unfortunately don't have time to test, I doubt it will work but you never know your luck, if you want to try it something may display on screen though there were errors in two files which I #defined or #ifdef out to get a build to complete.

I find myself too busy with other projects and all too often get sidetracked from curiosity to play with things I will not be able to put much effort into.  I drop the tar file here in case anyone else wants to try looking at this....

I think README and compileinfo are only textfiles with notes to read, and run build_kdrive.sh should build a Xserver in 10minutes, I may try testing in a few weeks if I get free time but doubt I go further.

Anyways, you give good support to my questions and seem to be create an interesting project, I wish you good luck and hope best for you. 

[tilkX.zip](https://github.com/vvaltchev/tilck/files/9781370/tilkX.zip)

**Comment by datamaskinen** [comment]
maybe this one is more easy to build
https://github.com/wizbright/waybox


On Fri, Oct 14, 2022 at 2:04 AM mung-kie ***@***.***> wrote:

> I am afraid I got a bit busy and doubt I can do anything of use, so
> appologise as I may have wasted your time.
> I managed to find some time this evening so have compiled the Xfbdev
> kdrive server using the tilck toolchain.
>
> But unfortunately don't have time to test I doubt it will work but you
> never know your luck if you want to try it something may display on screen
> though there were errors in two files which I #defined or #ifdef out to get
> a build to complete.
>
> I find myself too busy with other projects and all too often get
> sidetracked from curiosity to play with things I will not be able to put
> much effort into. I drop the tar file here in case anyone else wants to try
> looking at this....
>
> Anyways, you give good support to my questions and seem to be create an
> interesting project, I wish you good luck and hope best for you.
>
> tilkX.zip <https://github.com/vvaltchev/tilck/files/9781370/tilkX.zip>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vvaltchev/tilck/discussions/81#discussioncomment-3874506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADGCQMJR25XTDBNPLTLEYLTWDCPSLANCNFSM5BT2NVHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 


*Martin Lindkvist*

**Reply by mung-kie** [reply]
I find your comment irrelevent and your private messages homophobic and harassing.
If you do not stop harassing me I will have to report you.

**Reply by vvaltchev** [reply]
> I find your comment irrelevent and your private messages homophobic and harassing. If you do not stop harassing me I will have to report you.

@mung-kie  Maybe you mistakenly post this message in the wrong place? Please, note that any form of SPAM is not welcome here.

**Comment by mung-kie** [comment]
Okay, I found an hour free this morning so did quick test and find the zip file I posted has major error: there is no -static flag for compiler, so I recompile and put in sysroot and do very quick test (cannot run fb in qemu in vmware linuxmint).

But test shows fb is irrellevant as there is not even virtual terminal available yet.

SO I am sorry to say I leave this here, there is not enough kernel infrastructure available yet.

I maybe return in year or so and play with your kernel again.

I say thank you for your effort, you do good work, and I hope make good progress.

Good luck.

<img width="854" alt="tilckssxfbdev" src="https://user-images.githubusercontent.com/114813840/195991039-2797fc40-346d-4ad9-8474-40632372b9e0.png">

**Reply by vvaltchev** [reply]
> SO I am sorry to say I leave this here, there is not enough kernel infrastructure available yet.

That is what I was trying to tell you from the beginning. 

Please take a look at other OSDEV projects. There is a long list here: https://wiki.osdev.org/Projects
There is also a list with the most advanced projects: https://wiki.osdev.org/Notable_Projects
Many of them have great interest in running graphical applications.

Tilck won't have support for an X server any time soon, simply because it's **not** part of the project's goals. Tilck aims at becoming something like a real-time OS for embedded systems. On those systems, the (minimal) OS controls the hardware and runs a small and predefined set of applications (services), connecting with the outside world via network. It's **not meant** to be used directly by end users. Think about powerful micro-controllers used inside your car or to control robots in the industry.

Take a look here to learn more about that: https://en.wikipedia.org/wiki/Real-time_operating_system
One of the most successful OSs like that is FreeRTOS: https://github.com/FreeRTOS/FreeRTOS
Tilck will have some of the features of Embedded Linux, so it won't be exactly like FreeRTOS, but certainly **it won't become a desktop operating system**. Wishfully, it will become a hard real-time OS able to run _some_ simple Linux applications directly.

In conclusion, probably the most realistic answer to "_when Tilck will support X?_" is **never**.

**Comment by hatonthecat** [comment]
This is interesting. I like both deterministic OSes and linux. Symbian did both in a very small space. I am curious if the kernel could be paired to a UI layer so it has a UI but still works in a real-time way. https://en.wikipedia.org/wiki/Symbian#Symbian_UI_variants,_platforms https://github.com/SymbianSource

**Reply by vvaltchev** [reply]
@hatonthecat Well, that's certainly possible. Even today the framebuffer could be used for graphical applications. Once Tilck has the ability to run on typical embedded SoC (with networking etc.), adding support for touchscreen displays will be very useful. The idea is Tilck to be used on devices instead of desktop machines. 

Currently, my other project https://github.com/vvaltchev/tfblib has native integration with Tilck and it's already used on Linux for phones (by the postmarketOS distro). Once the kernel has the rest of required features, it will be fun to extend that library with support for UI widgets or use something else. Just, now it's way too early for that.

**Reply by cestith** [reply]
tfblib, SDL, or OpenGL to a framebuffer would be more than enough graphics for most small systems. I think that’s a wise choice of target for your goals.
